src/target.rs

```diff
@@ -0,0 +1,415 @@
+            (_, gnu) if gnu.starts_with("gnu") => {
+                let abi = gnu.strip_prefix("gnu").unwrap();
+                // To handle gnu_ilp32
+                let abi = abi.strip_prefix("_").unwrap_or(abi);
+                ("gnu", abi)
+            }
+            (_, musl) if musl.starts_with("musl") => {
+                let abi = musl.strip_prefix("musl").unwrap();
+                ("musl", abi)
+            }
+            (_, "softfloat") => ("", "softfloat"),
+            (_, "qnx700") => ("nto70", ""),
+            (_, "qnx710") => ("nto71", ""),
+            (_, "elf") => ("", ""),
+            (_, "none") => ("", ""),
+            (_, "eabi") => ("", "eabi"),
+            (_, "eabihf") => ("", "eabihf"),
+            (_, "uclibceabi") => ("uclibc", "eabi"),
+            (_, "uclibceabihf") => ("uclibc", "eabihf"),
+            (_, "newlibeabihf") => ("newlib", "eabihf"),
+            // General fallback
+            (_, env) => (env, ""),
+        };
+
+        // FIXME(rustc): The name has `pc` as the vendor, but it's not set?
+        if os == "nto" {
+            vendor = "unknown";
+        }
+
+        // FIXME(rustc): Vendor isn't set for these targets?
+        if vendor == "espressif" && os == "none" {
+            vendor = "unknown";
+        }
+
+        // Extra overrides for badly named targets.
+        match target {
+            // Actually simulator targets.
+            "i386-apple-ios" | "x86_64-apple-ios" | "x86_64-apple-tvos" => {
+                abi = "sim";
+            }
+            // Name should've contained `muslabi64`.
+            "mips64-openwrt-linux-musl" => {
+                abi = "abi64";
+            }
+            // Has no OS name
+            "avr-unknown-gnu-atmega328" => {
+                os = "none";
+                // FIXME(rustc): Should this specify `gnu` env?
+                env = "";
+                abi = "";
+            }
+            // Specifies abi even though not in name
+            "armv6-unknown-freebsd" | "armv7-unknown-freebsd" => {
+                // FIXME(rustc): Is it a bug that the env is set to "gnu" here?
+                env = "gnu";
+                abi = "eabihf";
+            }
+            // Specifies abi even though not in name
+            "armv7-unknown-linux-ohos" | "armv7-unknown-trusty" => {
+                abi = "eabi";
+            }
+            // FIXME(rustc): Is it a bug that the ABI isn't set here?
+            "armv7a-kmc-solid_asp3-eabi" | "armv7a-kmc-solid_asp3-eabihf" => {
+                abi = "";
+            }
+            // FIXME(rustc): Specifies "elf" ABI, is that correct?
+            "sparc-unknown-none-elf" => {
+                abi = "elf";
+            }
+            _ => {}
+        }
+
+        Ok(Target {
+            full_arch: full_arch.into(),
+            arch: arch.into(),
+            vendor: vendor.into(),
+            os: os.into(),
+            env: env.into(),
+            abi: abi.into(),
+        })
+    }
```

Comment:
So, I spent some time on actually implementing the target parsing, and as you can see with the test below, it _works_ now for all targets that `rustc` supports, but I wonder: Is this even the desired approach?

Another solution would be to run `rustc --print all-target-specs-json` at build time, and generate a static lookup table from target name to the target information. Such a table would need updating every time `rustc` adds a new target (when running outside of a build script), but that's honestly usually the case in `cc` anyhow, and at least this would be something automatic instead (and bug fixes in `rustc` to the target info would automatically get transfered to `cc`).

@NobodyXu wdyt?

Reply:
Doing it at build time is a bit hard:
 - IIRC `rustc --print all-target-specs-json` is only available on nightly
 - cc has msrv 1.63
 - cc can be used outside of build-script, and it's not guaranteed to have `rustc`

> Such a table would need updating every time rustc adds a new target (when running outside of a build script)

Yes we have a CI that run the generation and create PR if there's any update.

Reply:
Doing it at build time is a bit hard:
 - IIRC `rustc --print all-target-specs-json` is only available on nightly
 - cc has msrv 1.63
 - cc can be used outside of build-script, and it's not guaranteed to have `rustc`

> Such a table would need updating every time rustc adds a new target (when running outside of a build script)

Yes we have a CI that run the generation and create PR if there's any update.

Reply:
Okay, I'll try to implement that.

It does carry some problems, I've identified at least the following special cases:
1. A target triple is unsupported by `rustc` (e.g. because it is a custom / bare metal target, or because it is in the process of being upstreamed).
2. An outdated `cc` version that contains outdated data / the user is using an outdated version of `rustc` compared to `cc`'s data.

---

To handle the first case, we can read `CARGO_CFG_*` when in a build script, but we'll have to error when outside.

To solve this, it might make sense for `cc` to "vendor" target information for common bare metal targets, for example [the ones for the Arduino](https://github.com/Rahix/avr-hal/tree/main/avr-specs)? Though we could also declare it as not supported, or say that the user should use the `rustc` target, in the case of Arduino, as `.target("avr-unknown-gnu-atmega328")`?

(Actually, why _isn't_ this target information just upstreamed into `rustc`? That'd both be easier for us, `rustc`, the maintainers of `avr-hal`, and for the user?)

---

The second case is trickier, but I _think_ we want to use `CARGO_CFG_*` when available, and fall back to the static data in `cc` when not? That is, we prefer `rustc`'s view of the world, even though it may be outdated and incorrect.

As an example, `armv7a-kmc-solid_asp3-eabihf` has `eabihf` in the name, but does not specify `target_abi = "eabihf"`. Let's assume that it should (I'll open a PR later to `rustc` to figure out if that's true), and that it gets fixed in Rust 1.84.

Next, let's assume two versions of `cc` had been generated, `v1.1.25` with `rustc +1.81`, and `v1.1.26` with `rustc +1.84`. The first version would assume `target_abi = ""`, and the second would assume `target_abi = "eabihf"`.

This would make `cc v1.1.25` have the wrong behaviour when using `rustc +1.84`.

If we instead read the value from `CARGO_CFG_TARGET_ABI`, then we'd think the same as `rustc`, i.e. that it is `target_abi = ""` when using `rustc +1.81`, and `target_abi = "eabihf"` when using `rustc +1.84`.

This is obviously still incorrect, e.g. the user could be using `cc v1.1.26` and still get the wrong `target_abi = ""`, but at least the user doesn't get differing behaviour between things generated by `rustc`, and things generated by the C compiler that `cc` calls?

Honestly not sure here.

Reply:
My thoughts is that:
 - respect CARGO_CFG env if set, since it should be correct and up to date. If it is not, user can update their cargo (if people want cc to ignore it, we could expose an env/fn to ignore it, but I won't do it unless any user actually wants it).
 - otherwise fallback to data extracted from rustc at build time. If it is too old, then user should either upgrade cc or rustc/cargo
 - for non-official targets, using CARGO_CFG is preferred, and we can add popular non-official targets to gen-target-info (it's easier to update than directly hardcoding it in cc).